src/algebra/module/projective.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2021 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard
+-/
+
+import algebra.module.basic
+import linear_algebra.finsupp
+import linear_algebra.basis
+
+/-!
+
+# Projective modules
+
+This file contains a definition of a projective module, the proof that
+our definition is equivalent to a lifting property, and the
+proof that all free modules are projective.
+
+## Main definitions
+
+Let `R` be a ring (or a semiring) and let `M` be an `R`-module (or a semimodule).
+
+* `is_projective R M` : the proposition saying that `M` is a projective `R`-module.
+
+## Main theorems
+
+* `is_projective.lifting_property` : a map from a projective module can be lifted along
+  a surjection.
+
+* `is_projective.of_lifting_property` : If for all R-module surjections `A →ₗ B`, all
+  maps `M →ₗ B` lift to `M →ₗ A`, then `M` is projective.
+
+* `is_projective.of_free` : Free modules are projective
+
+## Implementation notes
+
+The actual definition of projective we use is that the natural R-module map
+from the free R-module on the type M down to M splits. This is more convenient
+than certain other definitions which involve quantifying over universes,
+and also universe-polymorphic (the ring and module can be in different universes).
+
+Everything works for semirings and semimodules except that apparently
+we don't have free semimodules, so here we stick to rings.
+
+`is_projective.of_lifting_property` is not universe polymorphic.
+
```

Comment:
```suggestion
```
since this is no longer accurate. I have no idea whether the stronger statement with `u` instead of `max u v` is even true, so would be inclined not to mention it.